### PR TITLE
Dash app for displaying an interactive eigenvalue plot.

### PR DESCRIPTION
**Pull request by sandertyu**

### CODE REVIEW

File: assets/styles.css [hunk 1]
```diff
@@ -0,0 +1,18 @@
+html {
+	background-color: grey;
+	background-image: url("/assets/bike-wall.png");
```
Comment (moorepants) [inline assets/styles.css]: I'd prefer not committing any binary files to the repository. In general, it isn't good practice. What is this image?

File: bicycleparameters/io.py [hunk 2]
```diff
@@ -109,6 +109,9 @@ def remove_uncertainties(dictionary):
             except TypeError:
                 # this is the case if the value is a float
                 noUncert[k] = v
+            except AttributeError:
+                # this is the case if the value is a float
+                noUncert[k] = v
```
Comment (moorepants) [inline bicycleparameters/io.py]: We should address this in a separate pull request. The precise issue needs to be identified and a unit test will need to be added.

File: bicycleparameters/io.py [hunk 3]
```diff
@@ -211,4 +214,4 @@ def write_periods_to_file(pathToRawFile, mp):
     # write it to the file
     f = open(pathToRawFile, 'w')
     f.write(withPeriods)
-    f.close()
+    f.close()
```
Comment (moorepants) [inline bicycleparameters/io.py]: newlines should remain, even on the last line

File: bicycleparameters/main.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 import matplotlib.pyplot as plt
 from matplotlib.patches import Ellipse, Wedge
 from uncertainties import unumpy
-from dtk import control
+#from dtk import control
```
Comment (moorepants) [inline bicycleparameters/main.py]: If this is causing issues for you, you need to make sure dynamicisttoolkit is installed. Undo this comment.

File: bicycleparameters/main.py [hunk 5]
```diff
@@ -35,8 +35,8 @@ def __new__(cls, bicycleName, pathToData='.', forceRawCalc=False,
         try:
             pathToBicycle = os.path.join(pathToData, 'bicycles', bicycleName)
             if os.path.isdir(pathToBicycle) == True:
-                print("We have foundeth a directory named: " +
-                      "{0}.".format(pathToBicycle))
+                '''print("We have foundeth a directory named: " +
+                      "{0}.".format(pathToBicycle))'''
```
Comment (moorepants) [inline bicycleparameters/main.py]: Undo this comment. We can remove these notifications but they should be addressed in a separate pull request.

File: bicycleparameters/main.py [hunk 6]
```diff
@@ -113,7 +113,7 @@ def __init__(self, bicycleName, pathToData='.', forceRawCalc=False,
         isRawDataDir = 'RawData' in os.listdir(self.directory)
 
         if isRawDataDir:
-            print("Found the RawData directory:", rawDataDir)
+            #print("Found the RawData directory:", rawDataDir)
```
Comment (moorepants) [inline bicycleparameters/main.py]: undo

File: bicycleparameters/main.py [hunk 7]
```diff
@@ -136,9 +136,9 @@ def __init__(self, bicycleName, pathToData='.', forceRawCalc=False,
                   % self.bicycleName)
         elif not forceRawCalc and isBenchmark:
             # we already have what we need
-            stmt1 = "Looks like you've already got some parameters for %s, "
+            '''stmt1 = "Looks like you've already got some parameters for %s, "
             stmt2 = "use forceRawCalc to recalculate."
-            print((stmt1 + stmt2) % self.bicycleName)
+            print((stmt1 + stmt2) % self.bicycleName)'''
```
Comment (moorepants) [inline bicycleparameters/main.py]: undo

File: bicycleparameters/main.py [hunk 8]
```diff
@@ -479,7 +479,7 @@ def add_rider(self, riderName, reCalc=False, draw=False):
             self.riderName = riderName
             self.hasRider = True
 
-    def plot_bicycle_geometry(self, show=True, pendulum=True,
+    def plot_bicycle_geometry(self, show=False, pendulum=True,
```
Comment (moorepants) [inline bicycleparameters/main.py]: This should not be changed in main, but you can change it when you call the function.

File: bicycleparameters/main.py [hunk 9]
```diff
@@ -877,7 +877,7 @@ def plot_eigenvalues_vs_speed(self, speeds, fig=None, generic=False,
         figsize = [figwidth, figwidth * goldenMean]
         params = {#'backend': 'ps',
             'axes.labelsize': 8,
-            'text.fontsize': 10,
+            #'text.fontsize': 10, # removed because out of date
```
Comment (moorepants) [inline bicycleparameters/main.py]: What are the consequences of this? Are the fontsizes now too large?

File: bicycleparameters/main.py [hunk 10]
```diff
@@ -1246,4 +1246,4 @@ def is_fork_split(mp):
         if key[:1] == 'S' or key[1:2] == 'S':
             forkIsSplit = True
 
-    return forkIsSplit
+    return forkIsSplit
```
Comment (moorepants) [inline bicycleparameters/main.py]: undo

File: cycle-app.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+import numpy as np
+
+###!!!Must ensure that current working directory is set to directory immediatly containing '\data\'!!!###
+###---------------------------------------------------------------------------------------------------###
+
+# Initializes bike data as Bicycle and retrieves its parameters
+
+
+def new_par(bike_name):
+    bike = bp.Bicycle(bike_name, pathToData=os.getcwd()+'\\data')
```
Comment (moorepants) [inline cycle-app.py]: I think that the data directory is installed with bicycle parameters. A better way to locate this, is relative to this file. But I think the cycle-app.py file needs to be moved inside the bicycleparameters/ directory first, so that it will be packaged along with the other files.

File: cycle-app.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+@app.callback([Output('geometry-plot', 'src'),
+               Output('eigen-plot', 'src')],
+              [Input('bike-dropdown', 'value'),
+               Input('wheel-table', 'data'),
+               Input('frame-table', 'data'),
+               Input('general-table', 'data')])
+def plot_update(value, x, y, z):
+    ctx = dash.callback_context
+    speeds = np.linspace(0, 10, num=100)
```
Comment (moorepants) [inline cycle-app.py]: It would be nice if the user could specify the minimum and maximum speeds via the GUI. If you do that, then the num parameter should be calculated such that there is data point every 0.1 m/s or so.

File: cycle-app.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+    wheelData = ctx.inputs.get('wheel-table.data')
+    frameData = ctx.inputs.get('frame-table.data')
+    genData = ctx.inputs.get('general-table.data')
+    newP = []
+    currentBike = bp.Bicycle(value, pathToData=os.getcwd()+'\\data')
```
Comment (moorepants) [inline cycle-app.py]: We are going to need a better way to package the data files alongside the app.

**Comment by moorepants** [comment]
The code is looking good! I left some comments.

**Comment by moorepants** [comment]
I think that we should do this for the app file and data files:

```
BicycleParameters/  # main repository
--bicycleparameters/  # main package
----cycle-app.py  # app file
----app-data/  # contains text files for the bicycles we want to provide as default bicycles in the app (copy those you want from BicycleParameters/data/ but don't delete them there)
```

So move `cycle-app.py` into `BicycleParameters/bicycleparameters/` and create a new, simpler data directory for the app.

This will allow the files to be packaged properly and installed.

Once you do this, you can access the data files in `cycle-app.py` with:

```python
import os
path_to_app_data = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'app-data')
```

This let's you reference relative to the file.

**Comment by moorepants** [comment]
I made some updates in the master branch. Here is how you update this branch:

1. Commit all current changes (or git stash them).
2. `git checkout master`
3. `git pull upstream master` this pull in my changes in master to your local master branch
4. `git checkout cycle-app` changes back to your branch for this PR
5. `git merge master` merges changes I made into your PR branch
6. `git push origin cycle-app` updates the PR

This is how you can update your feature branch if changes are made to the master branch.

File: assets/styles.css (hunk 1 above)
Comment (sandertyu) [reply assets/styles.css]: It's the photo I'm using for the background wallpaper; I made sure that it was copyright labeled for reuse. I don't know why it is specifically being tagged as binary, but I switched it out for a web url instead and removed the local files.

File: bicycleparameters/main.py (hunk 4 above)
Comment (sandertyu) [reply bicycleparameters/main.py]: I searched online for a dynamicstoolkit or dtk module but I wasn't able to find anything. It never gave me issues before, but my python versions have gotten a little messed up and that likely had something to do with it.

Import has been uncommented.

File: bicycleparameters/main.py (hunk 9 above)
Comment (sandertyu) [reply bicycleparameters/main.py]: We debugged that line a little while back. It's a method that isn't available for the matplotlib figures anymore and it was preventing plots from being drawn properly.

File: cycle-app.py (hunk 12 above)
Comment (sandertyu) [reply cycle-app.py]: Sure. We'll put that in the works.

File: bicycleparameters/main.py (hunk 4 above)
Comment (moorepants) [reply bicycleparameters/main.py]: How are you installing packages? pip or conda or other?

File: bicycleparameters/main.py (hunk 9 above)
Comment (moorepants) [reply bicycleparameters/main.py]: I remember that, but what is the effect of removing this?

**Comment by moorepants** [comment]
Rename `cycle-app.py` to `cycle_app.py`. Dashes are minus signs in Python so you can import the modules if needed. Standard is to make lower case with underscores in Python file names.

File: bicycleparameters/main.py (hunk 9 above)
Comment (sandertyu) [reply bicycleparameters/main.py]: It makes it impossible to draw plots. The error is this;

KeyError: 'text.fontsize is not a valid rc parameter (see rcParams.keys() for a list of valid parameters)'

File: bicycleparameters/main.py (hunk 4 above)
Comment (moorepants) [reply bicycleparameters/main.py]: You should be able to do `pip install DynamicistToolkit` then.

File: bicycleparameters/main.py (hunk 9 above)
Comment (moorepants) [reply bicycleparameters/main.py]: Sorry, I wasn't clear. I'd like to know if this is removed, do the font sizes change to values that change the plot outputs? If so, then you need to implement an alternative to set these font sizes correctly.

File: bicycleparameters/app-data/Crescendo/RawData/CrescendoMeasured.txt [hunk 14]
```diff
@@ -0,0 +1,50 @@
+lF = 0.301+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Crescendo/RawData/CrescendoMeasured.txt:1]: I don't think you need these "Measured" files for the app.

File: bicycleparameters/app-data/Fisher/RawData/FisherMeasured.txt [hunk 15]
```diff
@@ -0,0 +1,42 @@
+lF = 0.26425+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Fisher/RawData/FisherMeasured.txt:1]: remove

File: bicycleparameters/app-data/Pista/RawData/PistaMeasured.txt [hunk 16]
```diff
@@ -0,0 +1,42 @@
+lF = 0.297+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Pista/RawData/PistaMeasured.txt:1]: remove

File: bicycleparameters/app-data/Rigid/RawData/RigidMeasured.txt [hunk 17]
```diff
@@ -0,0 +1,90 @@
+dF = 21.0780313+/-0.001, 21.0859688+/-0.001, 21.0899375+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Rigid/RawData/RigidMeasured.txt:1]: remove

File: bicycleparameters/app-data/Silver/RawData/SilverMeasured.txt [hunk 18]
```diff
@@ -0,0 +1,8 @@
+lcs = 0.46
```
Comment (moorepants) [inline bicycleparameters/app-data/Silver/RawData/SilverMeasured.txt:1]: remove

File: bicycleparameters/app-data/Yellow/RawData/YellowMeasured.txt [hunk 19]
```diff
@@ -0,0 +1,50 @@
+lF = 0.3045+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Yellow/RawData/YellowMeasured.txt:1]: remove

File: bicycleparameters/app-data/Yellowrev/RawData/YellowrevMeasured.txt [hunk 20]
```diff
@@ -0,0 +1,50 @@
+lF = 0.3045+/-0.001
```
Comment (moorepants) [inline bicycleparameters/app-data/Yellowrev/RawData/YellowrevMeasured.txt:1]: remove

File: bicycleparameters/cycle_app.py [hunk 21]
```diff
@@ -0,0 +1,313 @@
+import dash
+import dash_core_components as dcc
+import dash_html_components as html
+import dash_table as tbl
+from dash.dependencies import Input, Output, State
+
+import bicycleparameters as bp
```
Comment (moorepants) [inline bicycleparameters/cycle_app.py:7]: This can be a relative import. I think this is what you want:

```python
from . import bicycleparameters as bp
```

File: bicycleparameters/app-data/Crescendo/RawData/CrescendoMeasured.txt (hunk 14 above)
Comment (sandertyu) [reply bicycleparameters/app-data/Crescendo/RawData/CrescendoMeasured.txt:1]: Okay. I went ahead and locally deleted all the RawData folders as well. They don't contain anything else.

File: bicycleparameters/cycle_app.py (hunk 21 above)
Comment (sandertyu) [reply bicycleparameters/cycle_app.py:7]: That's what I thought as well. However, when I use it, I receive the error message below. I looked around online and couldn't find a direct solution, but I did see some things mentioning sys module. Wasn't really sure how to proceed.

````
File "c:/Users/xyzasdf/Desktop/bike-lab/BicycleParameters/bicycleparameters/cycle_app.py", line 7, in <module>
    from . import bicycleparameters as bp
ImportError: attempted relative import with no known parent package

File: bicycleparameters/main.py (hunk 4 above)
Comment (sandertyu) [reply bicycleparameters/main.py]: Done. Big help.

File: bicycleparameters/main.py (hunk 9 above)
Comment (sandertyu) [reply bicycleparameters/main.py]: I do not know. With the line commented, the plot looks fine and I don't think I would want to change the font sizes anyways. When I try to run it without commenting out that line, it immediately fails and does not draw the plot at all. I can't assess the change. This is what I ran;
```
import bicycleparameters as bp
import numpy as np
import os

speed = np.linspace(0,10,num=100)
bike = bp.Bicycle('Benchmark', pathToData=os.getcwd()+'\\data')
bike.plot_eigenvalues_vs_speed(speed, show=True)
```

This is what I got
```
  File "C:\Users\xyzasdf\AppData\Local\Programs\Python\Python38\lib\site-packages\matplotlib\__init__.py", line 811, in __setitem__
    raise KeyError(
KeyError: 'text.fontsize is not a valid rc parameter (see rcParams.keys() for a list of valid parameters)'
```

File: bicycleparameters/cycle_app.py (hunk 21 above)
Comment (sandertyu) [reply bicycleparameters/cycle_app.py]: I've been at this for hours and still no use. I've been trying to do other sorts of imports from within the bicycleparameters directory and absolutely nothing is working. They all give similar "no parent package" errors. Here is one call made from within the interpreter which may shed some light, although I don't personally know entirely how to interpret it.

```
C:\Users\xyzasdf\Desktop\bike-lab\BicycleParameters\bicycleparameters>__init__.py
Traceback (most recent call last):
  File "C:\Users\xyzasdf\Desktop\bike-lab\BicycleParameters\bicycleparameters\__init__.py", line 3, in <module>
    from .main import Bicycle
ImportError: attempted relative import with no known parent package
```

File: bicycleparameters/cycle_app.py [hunk 22]
```diff
@@ -0,0 +1,316 @@
+import dash
+import dash_core_components as dcc
+import dash_html_components as html
+import dash_table as tbl
+from dash.dependencies import Input, Output, State
+
+from . import bicycleparameters as bp
```
Comment (moorepants) [inline bicycleparameters/cycle_app.py:7]: from .main import Bicycle

Comment (moorepants) [reply bicycleparameters/cycle_app.py:7]:
```python
from .io import remove_uncertainties
```

File: bicycleparameters/main.py [hunk 23]
```diff
@@ -879,7 +879,7 @@ def plot_eigenvalues_vs_speed(self, speeds, fig=None, generic=False,
         figsize = [figwidth, figwidth * goldenMean]
         params = {#'backend': 'ps',
             'axes.labelsize': 8,
-            'text.fontsize': 10,
+            'text.fontsize': 10, # removed because out of date
             'legend.fontsize': 8,
             'xtick.labelsize': 6,
             'ytick.labelsize': 6,
```
Comment (moorepants) [inline bicycleparameters/main.py:885]:
```python
try:
    plt.rcParams.update(params)
except KeyError:
    del params['text.fontsize']
    plt.rcParams.update(params)
```

File: bicycleparameters/cycle_app.py (hunk 21 above)
Comment (moorepants) [reply bicycleparameters/cycle_app.py:7]: Actually I think it just needs to be what you have: `import bicycleparameters as bp`. This file shouldn't use relative imports because it isn't imported.

Comment (moorepants) [reply bicycleparameters/cycle_app.py:7]: Sorry, I may have wasted a bunch of your time with the first statement. I haven't packaged in this precise way before so I assumed wrong.

Comment (moorepants) [reply bicycleparameters/cycle_app.py:7]: You may have to run it from the main directory though:

```
$ pwd
/path/to/BicycleParameters
$ python bicycleparameters/cycle_app.py
```

**Comment by moorepants** [comment]
I'm running the latest commit and I get this:

```
moorepants@garuda:BicycleParameters(cycle-app)$ python bicycleparameters/cycle_app.py 
Running on http://127.0.0.1:8050/
Debugger PIN: 689-724-440
 * Serving Flask app "cycle_app" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
Running on http://127.0.0.1:8050/
Debugger PIN: 782-612-664
Are you nuts?! Make a directory called 'Benchmark' with basic data
for your bicycle in this directory: '/home/moorepants/src/BicycleParameters/bicycleparameters/app-data'. Then I can actually
create a bicycle object. You may either need to change to the
correct directory or reset the pathToData argument.
Are you nuts?! Make a directory called 'Benchmark' with basic data
for your bicycle in this directory: '/home/moorepants/src/BicycleParameters/bicycleparameters/app-data'. Then I can actually
create a bicycle object. You may either need to change to the
correct directory or reset the pathToData argument.Are you nuts?! Make a directory called 'Benchmark' with basic data
for your bicycle in this directory: '/home/moorepants/src/BicycleParameters/bicycleparameters/app-data'. Then I can actually
create a bicycle object. You may either need to change to the
correct directory or reset the pathToData argument.
```

Doesn't look like its finding the data.

**Comment by moorepants** [comment]
I see, it is missing the `bicycles` directory.

File: bicycleparameters/app-data/Benchmark/Parameters/BenchmarkBenchmark.txt [hunk 24]
```diff
@@ -0,0 +1,26 @@
+w = 1.02+/-0.0
```
Comment (moorepants) [inline bicycleparameters/app-data/Benchmark/Parameters/BenchmarkBenchmark.txt:1]: The data files need to be in:

`bicycleparameters/app-data/bicycles/Benchmark/Parameters/BenchmarkBenchmark.txt`

You are missing `/bicycles/`.

**Comment by moorepants** [comment]
I just figured out this import issue a bit more. From this: https://stackoverflow.com/questions/14132789/relative-imports-for-the-billionth-time

Here is what you need to do:

```
$ cd /path/to/BicycleParameters
$ python -m bicycleparameters.cycle_app
```

That should work!

File: bicycleparameters/app-data/Benchmark/Parameters/BenchmarkBenchmark.txt (hunk 24 above)
Comment (sandertyu) [reply bicycleparameters/app-data/Benchmark/Parameters/BenchmarkBenchmark.txt]: You're right. I updated the file structure. Let me know if there are any other problems with that. I'm still unable to run cycle_app.py from that directory and test for these issues. I ran it using as similar a method to you as I could, but still no dice.

```
xyzasdf@DESKTOP-GK4SQ6I MINGW64 ~/Desktop/bike-lab/BicycleParameters (cycle-app)
$ py bicycleparameters/cycle_app.py                                                                                     Traceback (most recent call last):
  File "bicycleparameters/cycle_app.py", line 7, in <module>
    import bicycleparameters as bp
ModuleNotFoundError: No module named 'bicycleparameters'
```

**Comment by sandertyu** [comment]
> I just figured out this import issue a bit more. From this: https://stackoverflow.com/questions/14132789/relative-imports-for-the-billionth-time
> 
> Here is what you need to do:
> 
> ```
> $ cd /path/to/BicycleParameters
> $ python -m bicycleparameters.cycle_app
> ```
> 
> That should work!

**Comment by moorepants** [comment]
The app is running for me! Great :smile:

**Comment by moorepants** [comment]
Can you reorder the elements in the app so that its:

- Title
- Figures (side by side like you have them)
- Speed sliders
- Table

I think putting the figures front and top is better.

**Comment by moorepants** [comment]
The bicycle selector could be above the figures, below the title.

**Comment by moorepants** [comment]
What about this slider https://dash.plotly.com/dash-core-components/rangeslider ?

You can make the full range -45 m/s to 45 m/s and the starting values 0 m/s to 10 m/s

**Comment by sandertyu** [comment]
> I just figured out this import issue a bit more. From this: https://stackoverflow.com/questions/14132789/relative-imports-for-the-billionth-time
> 
> Here is what you need to do:
> 
> ```
> $ cd /path/to/BicycleParameters
> $ python -m bicycleparameters.cycle_app
> ```
> 
> That should work!

This does allow me to boot up and run the app. However, after running it for some amount of time and playing with the values, I receive this error;

```
Exception ignored in: <function Image.__del__ at 0x00000256EB573790>
Traceback (most recent call last):
  File "C:\Users\xyzasdf\AppData\Local\Programs\Python\Python38\lib\tkinter\__init__.py", line 4014, in __del__
    self.tk.call('image', 'delete', self.name)
RuntimeError: main thread is not in main loop
Exception ignored in: <function Image.__del__ at 0x00000256EB573790>
Traceback (most recent call last):
  File "C:\Users\xyzasdf\AppData\Local\Programs\Python\Python38\lib\tkinter\__init__.py", line 4014, in __del__
    self.tk.call('image', 'delete', self.name)
RuntimeError: main thread is not in main loop
Tcl_AsyncDelete: async handler deleted by the wrong thread
```

It looks like it has something to do with our base encoding image method? Or it may be entirely on my end.

**Comment by moorepants** [comment]
Not sure about the new error. I am gonna make some supper and can have a look later.

**Comment by moorepants** [comment]
Not sure why tkinter is called at all...

**Comment by sandertyu** [comment]
> Can you reorder the elements in the app so that its:
> 
> * Title
> * Figures (side by side like you have them)
> * Speed sliders
> * Table
> 
> I think putting the figures front and top is better.

I went through the application and reordered all of these within the code. I've been meaning to go through and organize all of the elements through CSS. Ultimately, all of these elements should be forced to their position through CSS styling, I think.

**Comment by sandertyu** [comment]
> What about this slider https://dash.plotly.com/dash-core-components/rangeslider ?
> 
> You can make the full range -45 m/s to 45 m/s and the starting values 0 m/s to 10 m/s

I didn't know this was an option. This is much nicer and I will work on replacing the current sliders with this single one.

**Comment by sandertyu** [comment]
> Not sure about the new error. I am gonna make some supper and can have a look later.

Playing with it more, it seems to only happen when I move the speed sliders away from their default values and then back to the default. Let me implement the new slider, see if that resolves the issue, and then we can return to it from there.

**Comment by sandertyu** [comment]
I've updated it with the range-slider, and the issue is still happening. There doesn't seem to be any rhyme or reason, it just occurs after I play with it for a little. It replicates itself as I let it sit, and the app has essentially crashed at this point. I suspect it has to do with how we are running the application, given the RuntimeError. This is what shows in the terminal;

```
Exception ignored in: <function Image.__del__ at 0x000001B271694790>
Traceback (most recent call last):
  File "C:\Users\xyzasdf\AppData\Local\Programs\Python\Python38\lib\tkinter\__init__.py", line 4014, in __del__
    self.tk.call('image', 'delete', self.name)
RuntimeError: main thread is not in main loop
```

However, these error messages in my terminal are also accompanied by `Error for API call: _reload-hash` in the dash online error GUI. These basically happen anytime the application is not running, such as when I have the browser page open and I quit the program. I'll continue to look for solutions.

**Comment by moorepants** [comment]
If you are interested, here is an explanation of the newline at the end of a file: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

**Comment by moorepants** [comment]
tkinter is not "threadsafe" according to this: https://stackoverflow.com/questions/14694408/runtimeerror-main-thread-is-not-in-main-loop, but I don't know why tkinter is being called.

**Comment by moorepants** [comment]
This has something to do with calling savefig(): https://github.com/r9y9/deepvoice3_pytorch/issues/5

**Comment by moorepants** [comment]
https://stackoverflow.com/questions/50848127/python-plot-image-save-error

Trying adding:

```matplotlib.use('Agg')```

https://stackoverflow.com/questions/27147300/matplotlib-tcl-asyncdelete-async-handler-deleted-by-the-wrong-thread/29172195#29172195

**Comment by sandertyu** [comment]
> https://stackoverflow.com/questions/50848127/python-plot-image-save-error
> 
> Trying adding:
> 
> `matplotlib.use('Agg')`
> 
> https://stackoverflow.com/questions/27147300/matplotlib-tcl-asyncdelete-async-handler-deleted-by-the-wrong-thread/29172195#29172195

This solves the issue. I haven't done much testing at this point, but it also seems to be fixing the `index out of bounds` error that I was receiving upon startup. Priorities now are to double check that the callbacks are all playing nicely together, and implement proper formatting so all the important elements fit onto one page.

**Comment by moorepants** [comment]
If this is now as a base functional state, we could merge it. You could work on remaining issues in different PRs.

File: bicycleparameters/io.py (hunk 2 above)
Comment (moorepants) [reply bicycleparameters/io.py:114]: This is the one thing that hasn't been addressed. I'd like to know why this is needed.

File: bicycleparameters/io.py (hunk 3 above)
Comment (moorepants) [reply bicycleparameters/io.py:217]: This seems to still be here.

File: bicycleparameters/io.py (hunk 2 above)
Comment (sandertyu) [reply bicycleparameters/io.py:114]: The app works fine with it removed. The issue must have been on my end back when I made the edit. It's all solved and removed now.

File: bicycleparameters/main.py [hunk 25]
```diff
@@ -1248,4 +1252,4 @@ def is_fork_split(mp):
         if key[:1] == 'S' or key[1:2] == 'S':
             forkIsSplit = True
 
-    return forkIsSplit
+    return forkIsSplit
```
Comment (moorepants) [inline bicycleparameters/main.py:1255]: Sill one more newline!

**Comment by moorepants** [comment]
It's running very nicely for me! It all seems to function, I'm not getting any errors. Sweet!

So, fix that one new line and let's merge this for your very first pull request! Congrats.

This is super cool!

**Comment by moorepants** [comment]
Congratulations! Merging.

**Comment by sandertyu** [comment]
> Congratulations! Merging.

Awesome! I think I fixed the newline in main.py but it's really difficult to check. This is pretty cool, I'm happy with it.

**Comment by moorepants** [comment]
So, as you move forward I suggest that you try to fix small things one at a time, each with a new pull request. I typically create an issue for each "atomic" thing I want to address. And then make pull requests for each issue. You can see how we do it if you see the closed issues and pull requests on the skijumpapp:

https://gitlab.com/moorepants/skijumpdesign/-/merge_requests?scope=all&utf8=%E2%9C%93&state=merged

**Comment by sandertyu** [comment]
> So, as you move forward I suggest that you try to fix small things one at a time, each with a new pull request. I typically create an issue for each "atomic" thing I want to address. And then make pull requests for each issue. You can see how we do it if you see the closed issues and pull requests on the skijumpapp:
> 
> https://gitlab.com/moorepants/skijumpdesign/-/merge_requests?scope=all&utf8=%E2%9C%93&state=merged

Okay. So they are very short and precise with only a few commits to them. Am I generally free to create pull requests as I see fit?

**Comment by moorepants** [comment]
Yes, you can create them as needed.

You'll want to checkout master, git pull upstream master, so your master branch is up-to-date with this one. Then you will `git checkout -b my_new_pr` for a new branch. This will be the new thing you want to add or fix, which you will push and then make a new PR for.

**Comment by moorepants** [comment]
Here are some basic instructions: https://www.asmeurer.com/git-workflow/

The idea is to now incrementally improve it with new PRs that we merge. Each PR should address a specific thing. For example, you might make a PR to adjust the colors. Or one to add the background image. Or one to add a new bike. etc.

**Comment by sandertyu** [comment]
> Here are some basic instructions: https://www.asmeurer.com/git-workflow/
> 
> The idea is to now incrementally improve it with new PRs that we merge. Each PR should address a specific thing. For example, you might make a PR to adjust the colors. Or one to add the background image. Or one to add a new bike. etc.

Got it, will do. I'll probably update the callback flow next and then rework the element formatting, using 2 different pull requests.
